File: library/src/main/java/me/proxer/library/api/HeaderInterceptor.java
package me.proxer.library.api;

import me.proxer.library.util.ProxerUrls;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;

/**
 * @author Ruben Gees
 */
final class HeaderInterceptor implements Interceptor {

    private static final String API_KEY_HEADER = "proxer-api-key";
    private static final String HEADER_USER_AGENT = "User-Agent";

    private final String apiKey;
    private final String userAgent;

    HeaderInterceptor(@NotNull final String apiKey, @Nullable final String userAgent) {
        this.apiKey = apiKey;
        this.userAgent = userAgent;
    }

    @Override
    public Response intercept(final Chain chain) throws IOException {
        final Request oldRequest = chain.request();

        if (oldRequest.url().host().equals(ProxerUrls.apiBase().host())) {
            final Request.Builder newRequestBuilder = chain.request().newBuilder().addHeader(API_KEY_HEADER, apiKey);

            if (userAgent != null) {
                newRequestBuilder.addHeader(HEADER_USER_AGENT, userAgent);
            }

            return chain.proceed(newRequestBuilder.build());
        } else {
            return chain.proceed(oldRequest);
        }
    }
}

File: library/src/main/kotlin/me/proxer/library/api/HeaderInterceptor.kt
package me.proxer.library.api

import me.proxer.library.util.ProxerUrls
import okhttp3.Interceptor
import okhttp3.Response

/**
 * @author Ruben Gees
 */
internal class HeaderInterceptor(private val apiKey: String, private val userAgent: String) : Interceptor {

    private companion object {
        private const val API_KEY_HEADER = "proxer-api-key"
        private const val USER_AGENT_HEADER = "User-Agent"
    }

    override fun intercept(chain: Interceptor.Chain): Response {
        val oldRequest = chain.request()

        if (ProxerUrls.hasProxerHost(oldRequest.url())) {
            val newRequestBuilder = oldRequest.newBuilder()

            if (oldRequest.url().host() == ProxerUrls.apiBase.host()) {
                newRequestBuilder.header(API_KEY_HEADER, apiKey)
            }

            newRequestBuilder.header(USER_AGENT_HEADER, userAgent)

            return chain.proceed(newRequestBuilder.build())
        } else {
            throw IllegalArgumentException("Only use ProxerLib's OkHttp instance with Proxer.Me URLs!")
        }
    }
}
